File: c++/c15_3_prims_algorithm.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "util.h"
#include "c12_3_applications.h"

namespace {
	vector<vector<pair<ll, ll>>> prim(const vector<vector<pair<ll, ll>>>& adj) {
		if (adj.empty()) return {};
		vector<bool> processed(adj.size());
		vector<vector<pair<ll, ll>>> ret(adj.size());
		processed[0] = true;
		priority_queue<tuple<ll, ll, ll>, vector<tuple<ll, ll, ll>>, greater<>> q;
		for (auto [c, w] : adj[0])
			q.emplace(w, 0, c);
		while (!q.empty()) {
			auto [w, a, b] = q.top(); q.pop();
			if (processed[b])
				continue;
			processed[b] = true;
			ret[a].emplace_back(b, w);
			ret[b].emplace_back(a, w);
			for (auto [c, w] : adj[b])
				q.emplace(w, b, c);
		}
		return ret;
	}
}

TEST(C153PrimsAlgorithm, prim) {
	EXPECT_EQ(prim({}), (vector<vector<pair<ll, ll>>>{}));
	auto ret = prim({ {{1,3},{4,5}},{{0,3},{2,5},{4,6}},{{1,5},{3,9},{5,3}},{{2,9},{5,7}},{{0,5},{1,6},{5,2}},{{2,3},{3,7},{4,2}} });
	ll sum = 0;
	for (const auto& v : ret)
		for (auto [_, w] : v)
			sum += w;
	EXPECT_EQ(sum, 20 * 2);
	auto adj = delweight(ret);
	EXPECT_EQ(connected(adj).size(), 1);
	EXPECT_FALSE(finding_cycles(adj));
}

File: python/c15_3_prims_algorithm.py
import unittest
from heapq import heapify,heappush,heappop
from util import delweight
from c12_3_applications import connected,finding_cycles

def prim(adj):
    if len(adj) <= 1:
        return [[]]
    processed=[False]*len(adj)
    ret=[[] for _ in adj]
    processed[1]=True
    q=[(w,1,c) for c,w in adj[1]]
    heapify(q)
    while q:
        w,a,b=heappop(q)
        if processed[b]:
            continue
        processed[b]=True
        ret[a].append((b,w))
        ret[b].append((a,w))
        for c,w in adj[b]:
            heappush(q,(w,b,c))
    return ret

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual(prim([[]]),[[]])
        ret=prim([
            [],
            [(2,3),(5,5)],
            [(1,3),(3,5),(5,6)],
            [(2,5),(4,9),(6,3)],
            [(3,9),(6,7)],
            [(1,5),(2,6),(6,2)],
            [(3,3),(4,7),(5,2)]
        ])
        self.assertEqual(sum(w for r in ret for _,w in r),20*2)
        adj=delweight(ret)
        self.assertEqual(len([*connected(adj)]),1)
        self.assertFalse(finding_cycles(adj))

if __name__=='__main__':
    unittest.main()
